### PR TITLE
enh(auth): replace sso keycloak by openid connect

### PR DESCRIPTION
**Pull request by lpinsivy**
## Description

Replace Keycloak to generic OpenId Connect

Fix: #8096 #8097

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x (master)

<h2> How this pull request can be tested ? </h2>

See https://github.com/centreon/centreon-documentation/pull/350

## Checklist

- [ ] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: www/class/centreonAuth.SSO.class.php [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <?php
-/*
+/**
  * Copyright 2005-2019 Centreon
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:3]:
```suggestion
 * Copyright 2005-2020 Centreon
```

File: www/class/centreonAuth.SSO.class.php [hunk 2]
```diff
@@ -71,30 +71,45 @@ public function __construct(
...
                 }
             }
-        } elseif (isset($this->ssoOptions['keycloak_enable'])
-            && $this->ssoOptions['keycloak_enable'] == 1
-            && !empty($this->ssoOptions['keycloak_url'])
-            && !empty($this->ssoOptions['keycloak_redirect_url'])
-            && !empty($this->ssoOptions['keycloak_realm'])
-            && !empty($this->ssoOptions['keycloak_client_id'])
-            && !empty($this->ssoOptions['keycloak_client_secret'])
+        } elseif (isset($this->ssoOptions['openid_connect_enable'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:74]:
```suggestion
        } elseif (
            isset($this->ssoOptions['openid_connect_enable'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 3]
```diff
@@ -71,30 +71,46 @@ public function __construct(
...
+                $authUrl .= "&scope=" . $this->ssoOptions['openid_connect_scope'];
+            }
 
             $inputForce = filter_var(
                 $_POST['force'] ?? $_GET['force'] ?? null,
                 FILTER_SANITIZE_NUMBER_INT
             );
             if ((isset($inputForce) && $inputForce == 1)
-                || (isset($this->options_sso['keycloak_mode']) && $this->options_sso['keycloak_mode'] == 0)
+                || (isset($this->options_sso['openid_connect_mode']) && $this->options_sso['openid_connect_mode'] == 0)
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:113]:
```suggestion
                || (isset($this->options_sso['openid_connect_mode']) && (int) $this->options_sso['openid_connect_mode'] === 0)
```

File: www/class/centreonAuth.SSO.class.php [hunk 4]
```diff
@@ -168,23 +199,22 @@ protected function checkSsoClient()
                     return 1;
                 }
             }
-        } elseif (isset($this->options_sso['keycloak_enable'])
-            && $this->options_sso['keycloak_enable'] == 1
-            && isset($this->options_sso['keycloak_mode'])
-            && $this->options_sso['keycloak_mode'] == 1
+        } elseif (isset($this->options_sso['openid_connect_enable'])
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:202]:
```suggestion
        } elseif (
            isset($this->options_sso['openid_connect_enable'])
```

File: www/class/centreonAuth.SSO.class.php [hunk 5]
```diff
@@ -194,91 +224,126 @@ protected function checkSsoClient()
...
+        $data = [
             "client_id" => $clientId,
             "client_secret" => $clientSecret,
             "grant_type" => "authorization_code",
             "code" => $code,
             "redirect_uri" => $redirectUri
-        );
+        ];
+
         $ch = curl_init($url);
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:273]: could you add a try catch and execute the curl_close in the finally

File: www/class/centreonAuth.SSO.class.php [hunk 6]
```diff
@@ -194,91 +224,126 @@ protected function checkSsoClient()
...
-        $url = $base . "/realms/" . $realm . "/protocol/openid-connect/token/introspect";
-        $data = array(
+        $data = [
             "token" => $token,
             "client_id" => $clientId,
             "client_secret" => $clientSecret
-        );
+        ];
 
         $ch = curl_init($url);
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:309]: same

File: www/class/centreonAuth.SSO.class.php [hunk 7]
```diff
@@ -194,91 +224,126 @@ protected function checkSsoClient()
...
+
+        $result = curl_exec($ch);
+        curl_close($ch);
+
+        $resp = json_decode($result, true);
+        return $resp;
+    }
+
+    /**
+     * Get User Inofmration on OpenId Connect
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:327]:
```suggestion
     * Get User Information on OpenId Connect
```

File: www/class/centreonAuth.SSO.class.php [hunk 8]
```diff
@@ -194,91 +224,126 @@ protected function checkSsoClient()
...
+     *
+     * @param string $url        OpenId Connect Introspection Token Endpoint
+     * @param string $token      OpenId Connect Token Access
+     * @param bool   $verifyPeer disable SSL verify peer
+     *
+     * @return array
+     */
+    public function getOpenIdConnectUserInfo($url, $token, $verifyPeer)
+    {
+        $ch = curl_init($url);
```
Comment (sc979) [inline www/class/centreonAuth.SSO.class.php:337]: same

File: www/include/Administration/parameters/DB-Func.php [hunk 9]
```diff
@@ -713,56 +713,87 @@ function updateGeneralConfigData($gopt_id = null)
...
+        isset($ret["openid_connect_enable"]["yes"]) && $ret["openid_connect_enable"]["yes"] != null ? 1 : 0
     );
     updateOption(
         $pearDB,
-        "keycloak_mode",
-        isset($ret["keycloak_mode"]["keycloak_mode"]) && $ret["keycloak_mode"]["keycloak_mode"] != null
-            ? $pearDB->escape($ret["keycloak_mode"]["keycloak_mode"]) : 1
+        "openid_connect_mode",
+        isset($ret["openid_connect_mode"]["openid_connect_mode"])
+            && $ret["openid_connect_mode"]["openid_connect_mode"] != null
```
Comment (sc979) [inline www/include/Administration/parameters/DB-Func.php:723]: hard to read

File: www/include/Administration/parameters/general/form.ihtml [hunk 10]
```diff
@@ -105,18 +105,23 @@
...
-    <tr class="list_one"><td class="FormRowField"><img class="helpTooltip" name="keycloak_redirect_url">{$form.keycloak_redirect_url.label}</td><td class="FormRowValue">{$form.keycloak_redirect_url.html}</td></tr>
-    <tr class="list_one"><td class="FormRowField"><img class="helpTooltip" name="keycloak_realm">{$form.keycloak_realm.label}</td><td class="FormRowValue">{$form.keycloak_realm.html}</td></tr>
-    <tr class="list_one"><td class="FormRowField"><img class="helpTooltip" name="keycloak_client_id">{$form.keycloak_client_id.label}</td><td class="FormRowValue">{$form.keycloak_client_id.html}</td></tr>
-    <tr class="list_one"><td class="FormRowField"><img class="helpTooltip" name="keycloak_client_secret">{$form.keycloak_client_secret.label}</td><td class="FormRowValue">{$form.keycloak_client_secret.html}</td></tr>
+    <tr class="list_one"><td class="FormRowField"><img class="helpTooltip" name="openid_connect_enable">{$form.openid_connect_enable.label}</td><td class="FormRowValue">{$form.openid_connect_enable.html}</td></tr>
```
Comment (sc979) [inline www/include/Administration/parameters/general/form.ihtml:111]: for better readability, you could split lines by sub-tags at least by opening <td>

File: www/install/php/Update-20.10.0.php [hunk 11]
```diff
@@ -1 +1,94 @@
...
+ */
+try {
+    $pearDB->beginTransaction();
+
+    /*
+     * Move keycloak configuration to OpenId Connect one
+     */
+    $errorMessage = "Unable to move Keycloak configuration to OpenId Connect";
+    $result = $pearDB->query(
+        "SELECT * FROM options WHERE options.key IN ('keycloak_enable', 'keycloak_mode', 'keycloak_url',
```
Comment (sc979) [inline www/install/php/Update-20.10.0.php:39]:
```suggestion
        "SELECT * FROM options WHERE options.`key` IN ('keycloak_enable', 'keycloak_mode', 'keycloak_url',
```

File: www/install/php/Update-20.10.0.php [hunk 12]
```diff
@@ -1 +1,94 @@
...
+    foreach ($openIdConnect as $key => $value) {
+        if (!is_null($value)) {
+            $statement->bindValue(':key', $key, \PDO::PARAM_STR);
+            $statement->bindValue(':value', $value, \PDO::PARAM_STR);
+            $statement->execute();
+        }
+    }
+
+    $pearDB->query(
+        "DELETE FROM options WHERE options.key IN ('keycloak_enable', 'keycloak_mode', 'keycloak_url',
```
Comment (sc979) [inline www/install/php/Update-20.10.0.php:77]:
```suggestion
        "DELETE FROM options WHERE options.`key` IN ('keycloak_enable', 'keycloak_mode', 'keycloak_url',
```

File: www/class/centreonAuth.SSO.class.php (hunk 5 above)
Comment (lpinsivy) [reply www/class/centreonAuth.SSO.class.php:273]: cURL does not throw exceptions, it is a thin wrapper for the C library it binds to.

File: www/class/centreonAuth.SSO.class.php (hunk 6 above)
Comment (lpinsivy) [reply www/class/centreonAuth.SSO.class.php:309]: cURL does not throw exceptions, it is a thin wrapper for the C library it binds to.

File: www/class/centreonAuth.SSO.class.php (hunk 8 above)
Comment (lpinsivy) [reply www/class/centreonAuth.SSO.class.php:337]: cURL does not throw exceptions, it is a thin wrapper for the C library it binds to.

File: www/class/centreonAuth.SSO.class.php [hunk 13]
```diff
@@ -71,30 +72,50 @@ public function __construct(
...
+            $clientSecret = $this->ssoOptions['openid_connect_client_secret'];
+            $redirectNoEncode = $this->ssoOptions['openid_connect_redirect_url'];
+            $baseUrl = rtrim($this->ssoOptions['openid_connect_base_url'], "/");
+            $authEndpoint = $baseUrl . rtrim($this->ssoOptions['openid_connect_authorization_endpoint'], "/");
+            $tokenEndpoint = $baseUrl . rtrim($this->ssoOptions['openid_connect_token_endpoint'], "/");
+            $introspectionEndpoint = $baseUrl . rtrim($this->ssoOptions['openid_connect_introspection_endpoint'], "/");
+            if (
+                isset($this->ssoOptions['openid_connect_userinfo_endpoint'])
+                && $this->ssoOptions['openid_connect_userinfo_endpoint'] != ""
+            ) {
```
Comment (kduret) [inline www/class/centreonAuth.SSO.class.php:96]:
```suggestion
            if (!empty($this->ssoOptions['openid_connect_userinfo_endpoint'])) {
```

File: www/class/centreonAuth.SSO.class.php [hunk 14]
```diff
@@ -71,30 +72,50 @@ public function __construct(
...
+            $verifyPeer = $this->ssoOptions['openid_connect_verify_peer'];
 
             $redirect = urlencode($redirectNoEncode);
-            $authUrl = $base . "/realms/" . $realm . "/protocol/openid-connect/auth?client_id=" . $clientId
-                . "&response_type=code&redirect_uri=" . $redirect;
+            $authUrl = $authEndpoint . "?client_id=" . $clientId . "&response_type=code&redirect_uri=" . $redirect;
+            if (
+                isset($this->ssoOptions['openid_connect_scope'])
+                && $this->ssoOptions['openid_connect_scope'] != ""
+            ) {
```
Comment (kduret) [inline www/class/centreonAuth.SSO.class.php:107]:
```suggestion
            if (!empty($this->ssoOptions['openid_connect_scope'])) {
```

File: www/include/Administration/parameters/DB-Func.php [hunk 15]
```diff
@@ -713,56 +713,87 @@ function updateGeneralConfigData($gopt_id = null)
...
+            ? $pearDB->escape($ret["openid_connect_mode"]["openid_connect_mode"]) : 1
```
Comment (kduret) [inline www/include/Administration/parameters/DB-Func.php:724]:
```suggestion
        !empty($ret["openid_connect_mode"]["openid_connect_mode"])
          ? $pearDB->escape($ret["openid_connect_mode"]["openid_connect_mode"])
          : 1
```

File: www/install/php/Update-20.10.0-beta.1.php [hunk 16]
```diff
@@ -1 +1,94 @@
...
+ *
+ * For more information : contact@centreon.com
+ *
+ */
+
+include_once __DIR__ . "/../../class/centreonLog.class.php";
+$centreonLog = new CentreonLog();
+
+//error specific content
+$versionOfTheUpgrade = 'UPGRADE - 20.10.0 : ';
```
Comment (sc979) [inline www/install/php/Update-20.10.0-beta.1.php:26]:
```suggestion
$versionOfTheUpgrade = 'UPGRADE - 20.10.0-beta.1 : ';
```

File: www/class/centreonAuth.SSO.class.php [hunk 17]
```diff
@@ -103,17 +118,32 @@ public function __construct(
                 $_POST['code'] ?? $_GET['code'] ?? null,
                 FILTER_SANITIZE_STRING
             );
-            if (isset($inputCode)) {
-                $keyToken = $this->getKeycloakToken(
-                    $base,
-                    $realm,
+
+            if (isset($inputCode) && !is_null($inputCode) && $inputCode != '') {
```
Comment (loiclau) [inline www/class/centreonAuth.SSO.class.php:122]: !empty($inputCode)

**Review by sc979 — DISMISSED** [review]

**Review by loiclau — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Review by loiclau — APPROVED** [review]

**Pull request merged**